### DISCUSSION TITLE
Support bumpVersion with prerelease semver level

### DISCUSSION THREAD

**Discussion by alanrossx2** [renovatebot/renovate#15573] [Suggest an Idea]
Hi community,

I was interested in being able to support prerelease version increments on maven dependency updates. At my current organization, we lightly comply with semver and tend to do most of our development with prerelease versions which are shared between multiple different applications. A typical update we are looking to apply could look like this:

1. Before: 
```
<version>1.0.0-1</version>

<dependencies>
  <dependency>
    <groupId>org.postgresql</groupId>
    <artifactId>postgresql</artifactId>
    <version>42.3.4</version>
  </dependency>
</dependencies>
```
2. After:
```
<version>1.0.0-2</version>

<dependencies>
  <dependency>
    <groupId>org.postgresql</groupId>
    <artifactId>postgresql</artifactId>
    <version>42.3.5</version>
  </dependency>
</dependencies>
```

My understanding is that the `bumpVersion` feature handles this exact task. However, it only currently supports `major`, `minor` and `patch` increments according to the following: https://docs.renovatebot.com/configuration-options/#bumpversion

### My understanding of how the bumpVersion options are set

To understand if prerelease versions could be incremented, I took a quick read of the source code and noticed the following:
1. The bumpVersion config is provided here: https://github.com/renovatebot/renovate/blob/982cefff2b8a3c697fe52fd5e023580a0678322e/lib/config/options/index.ts#L1221
2. The logic to increment the version uses the npm `semver` dependency: https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/maven/update.ts#L87

Once the semver dependency is installed, the logic is straight forward for incrementing the versions:
```
$ npm install -g semver
.
.
$ semver -i major 1.0.0
2.0.0
$ semver -i minor 1.0.0
1.1.0
$ semver -i patch 1.0.0
1.0.1
```

### My understanding of how prerelease version increments could be added to renovate

The latest semver dependency (used by renovate today) already supports prerelease:
```
$ semver --help | grep "\-\-increment" -A 4
-i --increment [<level>]
        Increment a version by the specified level.  Level can
        be one of: major, minor, patch, premajor, preminor,
        prepatch, or prerelease.  Default level is 'patch'.
        Only one version may be specified.
$ semver -i prerelease 1.0.0-1
1.0.0-2
```

In terms of actually enabling `prerelease`, I suspect we can just add it as an allowed values to the config. I did not quite understand how all the config is processed from start to end (from `renovateOption` -> `globalConfig` -> `config.upgrades` -> `upgrade.bumpVersion`), but from my quick read I suspect the following change would be sufficient to enable the feature:
```
  {
    name: 'bumpVersion',
    description: 'Bump the version in the package file being updated.',
    type: 'string',
    allowedValues: ['major', 'minor', 'patch', 'prerelease'],
    supportedManagers: ['helmv3', 'npm', 'maven', 'sbt'],
  },
```

Overall, I don't know if this is a reasonable ask, especially in terms of how it affects non maven versions. I also don't believe this is a duplicated question. Let me know your thoughts.

**Comment by rarkins** [comment]
Thanks for your thoughtful discussion on the topic. I think this suggestion is OK and would be happy to receive a pull request for it - I do not see any downsides

**Reply by alanrossx2** [reply]
Thanks Rarkins.
I have opened the relevant PR here: https://github.com/renovatebot/renovate/pull/15626

I have tested that this feature does what I need manually on my own repos.
I did not know where the unit tests fit in for this specific feature. I placed one for only maven but if you feel there should be more let me know.
